engine/engine.go

```diff
@@ -228,7 +228,16 @@ func handleSolveEvents(startOpts *Config, ch chan *bkclient.SolveStatus) error {
 			for ev := range ch {
 				cleaned := *ev
 				cleaned.Vertexes = make([]*bkclient.Vertex, len(ev.Vertexes))
+
+				// Avoid log duplication on multi stage.
+				// See https://github.com/dagger/dagger/issues/4605
+				isAlreadyLogged := false
 				for i, v := range ev.Vertexes {
+					if v.Started == nil && v.Completed == nil {
```

Comment:
I'm not sure if this is correct, but the buildkit code is so convoluted and also lacking of any tests so I can't really tell either way.

I am not really sure what situation vertices show up with started and completed set to nil. I think it might be when they first show up? But if that's the case the fact that we skip every vertex in this batch of those received seems like it could be wrong. cc @aluzzardi @vito in case you have any more insight given all your recent adventures.

I wonder if we could instead use `v.Cached` to help us here? Any of the duplicated vertices should show up cached, but if we have already received a vertex event saying it was cached we shouldn't need to display it? I am just spitballing here not totally sure.

Reply:
Yeah, my understanding is that you'll see started/completed both as `nil` when vertexes in the graph show up for the first time and aren't able to run immediately because they're waiting on an input. A common example of this would be building a `Dockerfile` with a bunch of `RUN` steps. So I don't think this is a good heuristic for detecting duplicate output.

I wonder if this issue has to do with the fact that Publish/Export need to initiate a new Buildkit client connection every time, and they just send those events down the same `chan<- *bkclient.SolveStatus` used for all the other API calls. https://github.com/moby/buildkit/issues/1914 seems relevant here again.

Also the SDK is likely causing the container to evaluate before it even calls Publish, since the repro [calls `Container.Directory`](https://gist.github.com/kpenfound/0de9abd133344fb4e69f31255d5e96fa#file-main-go-L28) which forces the container to evaluate.

I also noticed the log output isn't showing the timestamp to nanosecond precision (even though the repro does `$(date +%s%N)` which should), so it's hard to tell if the lines are duplicated or if it's somehow running twice.

Don't have a silver bullet, but hopefully some of this helps.

Reply:
> I wonder if we could instead use v.Cached to help us here? Any of the duplicated vertices should show up cached, but if we have already received a vertex event saying it was cached we shouldn't need to display it? I am just spitballing here not totally sure.

@sipsma I did some tests and it's kind of funny because there's log skipped with `v.Cached` set to true.
I kept my fix and add a display when I skip a log to see `v.Cached` and here is my result:

```shell
#1 resolve image config for docker.io/library/alpine:latest
#1 DONE 1.9s

#1 resolve image config for docker.io/library/alpine:latest
skipped log! v.Cached: false, v.Name: exec sh -c echo LOOK AT MEEEEEEEEEEEEEE now:$(date +%s%N) arch:$(uname -m) pid:$(pidof pidof)
skipped log! v.Cached: false, v.Name: {"name":"creating dagger metadata","internal":true}
skipped log! v.Cached: false, v.Name: pull docker.io/library/alpine:latest
#1 DONE 2.1s

#2 importing cache manifest from dagger:10686922502337221602
#2 DONE 0.0s

#3 
#3 ...

#4 from alpine
#4 resolve docker.io/library/alpine:latest 0.2s done
#4 sha256:c41833b44d910632b415cd89a9cdaa4d62c9725dc56c99a7ddadafd6719960f9 3.26MB / 3.26MB 0.3s done
#4 extracting sha256:c41833b44d910632b415cd89a9cdaa4d62c9725dc56c99a7ddadafd6719960f9 0.1s done
#4 DONE 0.5s

#3 
#3 0.603 LOOK AT MEEEEEEEEEEEEEE now:1680229134 arch:aarch64 pid:15
skipped log! v.Cached: false, v.Name: copy / /
#3 ...

#5 exporting to image
#5 exporting layers 0.3s done
#5 exporting manifest sha256:ab136db9fc3de34af3ee7d31b13535bcccae4a0bb4732e15b7d4c758d92ed73e done
#5 exporting config sha256:4a62bc2a88f75f470cc91a5fb05444aecb9de97e0b85221da41b643761594d1c done
#5 pushing layers 2.4s done
#5 pushing manifest for docker.io/the0only0vasek/test:latest@sha256:ab136db9fc3de34af3ee7d31b13535bcccae4a0bb4732e15b7d4c758d92ed73e 0.4s done
#5 DONE 3.1s
```

Result are the same with a new dagger engine or when operations are cached.

I also agree with @vito, https://github.com/moby/buildkit/issues/1914 should fix this problem but for now it is not available.

I'm out of options for now, do you guys have any ideas of something we can do on Dagger side to fix this issue?